### PR TITLE
Adding parameter to keep proportions

### PR DESCRIPTION
**Pull request by rgritti**

### CODE REVIEW

**Comment by danpersa** [comment]
Isn't this doing exactly what the `longerEdgeResize` filter is doing?

**Comment by rgritti** [comment]
@danpersa partially.
`longerEdgeResize` inscribe the image in a square and the parameter is the length of the square. while with `resize` keeping proportions you can inscribe a picture in a rectangle

File: README.md [hunk 1]
```diff
@@ -55,7 +55,7 @@ Skrop provides a set of filters, which you can use within the routes:
 * **crop(width, height, type)** — crops the image to have the specified width and height the type can be "north", "south", "east" and "west"
 * **cropByHeight(height, type)** — crops the image to have the specified height
 * **cropByWidth(width, type)** — crops the image to have the specified width
-* **resize(width, height)** — resizes an image without preserving its aspect ratio
+* **resize(width, height, <keep proportions?>)** — resizes an image. Third parameter is optional: true to keep proportions, false to ignore proportions (default to true)
```
Comment (danpersa) [inline README.md]: I would rename proportions to aspect ratio... it's a more common term for images.

File: filters/resize.go [hunk 2]
```diff
@@ -2,18 +2,20 @@ package filters
...
 const (
 	ResizeName = "resize"
 )
 
 type resize struct {
-	width  int
-	height int
+	width      int
+	height     int
+	proportion bool
```
Comment (danpersa) [inline filters/resize.go]: I would rename `proportion` to `keepAspectRatio`. It's a common way to refer to this use case, when dealing with images.

File: filters/resize.go [hunk 3]
```diff
@@ -53,6 +77,16 @@ func (r *resize) CreateFilter(args []interface{}) (filters.Filter, error) {
...
 	}
 
+	if len(args) == 3 {
+		f.proportion, err = parse.EskipBoolArg(args[2])
+
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		f.proportion = true
```
Comment (danpersa) [inline filters/resize.go]: `keepAspectRatio` should be false by default, as it's true for the `longerEdgeResize` filter

File: parse/parsers.go [hunk 4]
```diff
@@ -35,4 +35,12 @@ func EskipStringArg(arg interface{}) (string, error) {
 	} else {
 		return "", filters.ErrInvalidFilterParameters
 	}
-}
\ No newline at end of file
+}
+
+func EskipBoolArg(arg interface{}) (bool, error) {
```
Comment (danpersa) [inline parse/parsers.go]: You should add tests for this function

**Comment by rgritti** [comment]
do you think is it better to have 2 different filters? `resize` or `resizeWithAspectRatio`?

File: filters/resize_test.go [hunk 5]
```diff
@@ -17,14 +17,40 @@ func TestResize_Name(t *testing.T) {
 	assert.Equal(t, "resize", r.Name())
 }
 
-func TestResize_CreateOptions(t *testing.T) {
```
Comment (danpersa) [inline filters/resize_test.go]: You should not delete this test, when `keepAspectRatio` is not specified, you should test that it defaults to false, and that the aspect ratio is not preserved.

**Comment by danpersa** [comment]
@rgritti I think one filter with a flag is ok. (We already have the `longerEdgeResize` which preserves the aspect ratio.

File: filters/resize.go [hunk 6]
```diff
@@ -78,13 +78,13 @@ func (r *resize) CreateFilter(args []interface{}) (filters.Filter, error) {
...
 	if len(args) == 3 {
-		f.proportion, err = parse.EskipBoolArg(args[2])
+		f.keepAspectRatio, err = parse.EskipBoolArg(args[2])
 
 		if err != nil {
 			return nil, err
 		}
 	} else {
-		f.proportion = true
+		f.keepAspectRatio = true
```
Comment (danpersa) [inline filters/resize.go]: Why do you think it should default to `true`? We have the `longerEdgeResize` filter which keeps the aspect ratio be default.

**Comment by rgritti** [comment]
@danpersa I solved the comments except for the default value of `keepAspectRatio`. `longerEdgeResize` it is just a particular case of the `resize` in which `Width` and `Height` have the same length.
if frontend wants to put an image in a box that is not squared, eg a rectangle, they cannot use `longerEdgeResize`

**Comment by danpersa** [comment]
👍

File: README.md [hunk 7]
```diff
@@ -55,7 +55,7 @@ Skrop provides a set of filters, which you can use within the routes:
 * **crop(width, height, type)** — crops the image to have the specified width and height the type can be "north", "south", "east" and "west"
 * **cropByHeight(height, type)** — crops the image to have the specified height
 * **cropByWidth(width, type)** — crops the image to have the specified width
-* **resize(width, height)** — resizes an image without preserving its aspect ratio
+* **resize(width, height, <keep aspect ration?>)** — resizes an image. Third parameter is optional: true to keep aspect ratio, false to ignore (default to true)
```
Comment (danpersa) [inline README.md]: `keep aspect ratio`, not `ration` :)

**Pull request merged**